redical_core/src/calendar.rs

```diff
@@ -161,6 +161,25 @@ impl Calendar {
 
         Ok(true)
     }
+
+    // Iterates through associated events and finds those that have their last occurrence between
+    // the from and until timestamps.
+    pub fn prune_events(&mut self, from: i64, until: i64) -> Result<HashMap<String, Box<Event>>, String> {
+        let mut pruned_events = HashMap::new();
+
+        let event_uids_to_prune = self.events.iter()
+            .filter(|(_, event)| event.is_last_occurrence_between(from, until).unwrap_or(false))
+            .map(|(uid, _)| uid.clone())
+            .collect::<Vec<_>>();
+
+        for uid in event_uids_to_prune.iter() {
+            if let Some(pruned_event) = self.remove_event(uid) {
+                pruned_events.insert(uid.to_string(), pruned_event);
+            }
+        }
```

Comment:
Some small tweaks to make it more deliberate 🙂 

```suggestion
        let event_uids_to_prune = self.events.iter()
            .filter(|(_, event)| event.is_last_occurrence_between(from, until).unwrap_or(false))
            .map(|(uid, _)| uid.to_string())
            .collect::<Vec<String>>();

        for uid in event_uids_to_prune {
            if let Some(pruned_event) = self.remove_event(&uid) {
                pruned_events.insert(uid, pruned_event);
            }
        }
```

Reply:
I appreciate your approach here too, rather than wastefully cloning all the events you borrow them and create a `Vec<String>` of UIDs to remove which is slightly more computationally expensive but much more memory efficient which is 👍 with huge calendars!

Reply:
Nice one! Will update this too, makes sense :)